src/test/java/games/strategy/engine/GameEngineUtilsTest.java

```diff
@@ -0,0 +1,37 @@
+package games.strategy.engine;
+
+import static games.strategy.engine.GameEngineUtils.isEngineCompatibleWithEngine;
+import static games.strategy.engine.GameEngineUtils.isEngineCompatibleWithMap;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+
+import games.strategy.util.Version;
+
+public final class GameEngineUtilsTest {
+  @Test
+  public void testIsEngineCompatibleWithEngine() {
+    assertTrue(isEngineCompatibleWithEngine(new Version(1, 9), new Version(1, 9)));
```

Comment:
It may be easier to add more test cases to be more exhaustive by using a lambda foreach loop as shown below:
```
Arrays.stream(
  new Version(1,9),
  new Version(1,9,0,1))
.forEach(compatibleValue)) {
 assertTrue(isEngineCompatibleWithEngine(new Version(1, 9), compatibleValue);
}
```
Then you could also verify yet more cases to go with that. For example it does seem like we should have some `assertTrue` cases with a major version of `2` to really be sure we have a contrast with the assertFalse cases.

Reply:
> Null comparsion checks look to be missing. If we fail to deserialize Version properly, it can be null, knowing the behavior of a null vs non-null would be useful.

As discussed elsewhere, without any clarification in the Javadocs, or the presence of a `@Nullable` annotation, callers should assume passing a `null` argument will result in an NPE.  Adding tests like `assertThrows(NullPointerException.class, obj.someMethod(null))` becomes quite a burden.  I worked on a codebase once where we did this, and it turned out that fully 50% of the unit tests were NPE checks.  A few years later, after we switched to a compiler that employed null analysis, we realized just how little value these tests were providing. :smile:

I would argue that a "deserialization failure" should be handled close to the point of failure instead of propagating `null` throughout the system and trying to come up with some rules for handling it.  In that case, raising an NPE would help us detect our "failure" to handle the failure.

> A secondary suggestion

I'll update the tests appropriately.

Reply:
Well said regarding null and its handling, I agree. As a side note, we have some work to do somewhere to fix the `Version` deserialization case.

@ssoloff  I'm curious, was your team also able to get rid of `Optionals` in the same way you removed the tests?

Reply:
> Well said regarding null and its handling, I agree.

You actually suggested we do that a couple of months ago.  I'm just riding your coat-tails. :smile:

>  I'm curious, was your team also able to get rid of Optionals in the same way you removed the tests?

That particular codebase was Java 6/7, so `Optional` wasn't in use.  When we had to represent "optional" data back then, we tended to use Null Objects instead of `null` whenever possible.